File: packages/soulseek_protocol/lib/src/connection/socket_manager.dart
import 'dart:async';
import 'dart:io';
import 'dart:typed_data';

import '../messages/buffer.dart';

enum SocketState { disconnected, connecting, connected }

enum SocketErrorType {
  none,
  connectionRefused,
  dnsFailure,
  timeout,
  tlsError,
  networkUnreachable,
  connectionReset,
  unknown,
}

class SocketStateChanged {
  final SocketState state;
  final SocketErrorType errorType;
  final String? errorMessage;

  SocketStateChanged({
    required this.state,
    this.errorType = SocketErrorType.none,
    this.errorMessage,
  });
}

class MessageReceived {
  final SoulseekMessage message;
  MessageReceived(this.message);
}

class SocketManager {
  Socket? _socket;
  SocketState _state = SocketState.disconnected;
  StreamSubscription? _dataSubscription;
  StreamSubscription? _errorSubscription;
  StreamSubscription? _doneSubscription;

  final _messageController = StreamController<SoulseekMessage>.broadcast();
  final _stateController = StreamController<SocketStateChanged>.broadcast();

  final List<int> _buffer = [];
  static const int _headerSize = 8;
  Timer? _connectTimeout;

  SocketState get state => _state;
  Stream<SoulseekMessage> get messages => _messageController.stream;
  Stream<SocketStateChanged> get stateChanges => _stateController.stream;

  void dispose() {
    _messageController.close();
    _stateController.close();
    _cleanup();
  }

  Future<void> connect(String host, int port, {Duration timeout = const Duration(seconds: 10)}) async {
    if (_state == SocketState.connecting || _state == SocketState.connected) {
      await disconnect();
    }

    _setState(SocketState.connecting);

    try {
      final addresses = await _resolveHost(host);
      if (addresses.isEmpty) {
        _setState(SocketState.disconnected, SocketErrorType.dnsFailure, 'Could not resolve host: $host');
        return;
      }

      _socket = await Socket.connect(
        addresses.first,
        port,
        timeout: timeout,
      );

      _socket!.setOption(SocketOption.tcpNoDelay, true);

      _dataSubscription = _socket!.listen(
        _onData,
        onError: (error) => _onError(error, StackTrace.current),
        onDone: _onDone,
        cancelOnError: false,
      );

      _setState(SocketState.connected);
    } on SocketException catch (e) {
      _setState(SocketState.disconnected, _classifySocketError(e), e.message);
    } on TimeoutException {
      _setState(SocketState.disconnected, SocketErrorType.timeout, 'Connection timed out after $timeout');
    } catch (e) {
      _setState(SocketState.disconnected, SocketErrorType.unknown, e.toString());
    }
  }

  Future<List<InternetAddress>> _resolveHost(String host) async {
    try {
      final results = await InternetAddress.lookup(host, type: InternetAddressType.any);
      // Filter IPv4 first, fall back to IPv6
      final ipv4 = results.where((a) => a.type == InternetAddressType.IPv4).toList();
      final ipv6 = results.where((a) => a.type == InternetAddressType.IPv6).toList();

      if (ipv4.isNotEmpty) return ipv4;
      if (ipv6.isNotEmpty) return ipv6;
      return results;
    } catch (_) {
      return [];
    }
  }

  Future<void> disconnect() async {
    _connectTimeout?.cancel();
    _connectTimeout = null;
    _buffer.clear();
    await _dataSubscription?.cancel();
    await _errorSubscription?.cancel();
    await _doneSubscription?.cancel();
    _dataSubscription = null;
    _errorSubscription = null;
    _doneSubscription = null;
    await _socket?.close();
    _socket = null;
    _setState(SocketState.disconnected);
  }

  void sendMessage(SoulseekMessage message) {
    sendRaw(message.code, message.payload);
  }

  void sendRaw(int code, Uint8List payload) {
    if (_socket == null || _state != SocketState.connected) {
      throw SocketManagerException('Not connected');
    }
    final data = SoulseekMessage.encode(code, payload);
    _socket!.add(data);
  }

  void _onData(Uint8List data) {
    _buffer.addAll(data);
    _tryParseMessages();
  }

  void _tryParseMessages() {
    while (true) {
      if (_buffer.length < _headerSize) break;

      final lengthBytes = Uint8List.sublistView(Uint8List.fromList(_buffer.sublist(0, 4)));
      final totalLength = ByteData.view(lengthBytes.buffer).getUint32(0, Endian.little);

      final frameSize = 4 + totalLength; // length field + message
      if (_buffer.length < frameSize) break;

      final frame = Uint8List.fromList(_buffer.sublist(0, frameSize));
      _buffer.removeRange(0, frameSize);

      try {
        final message = SoulseekMessage.parse(frame);
        _messageController.add(message);
      } catch (e) {
        // Skip malformed frames
        continue;
      }
    }
  }

  void _onError(Object error, StackTrace stack) {
    _setState(SocketState.disconnected, SocketErrorType.connectionReset, error.toString());
  }

  void _onDone() {
    _setState(SocketState.disconnected, SocketErrorType.connectionReset, 'Connection closed');
    _cleanup();
  }

  void _cleanup() {
    _buffer.clear();
    _dataSubscription?.cancel();
    _errorSubscription?.cancel();
    _doneSubscription?.cancel();
    _dataSubscription = null;
    _errorSubscription = null;
    _doneSubscription = null;
    _socket = null;
  }

  void _setState(SocketState state, [SocketErrorType errorType = SocketErrorType.none, String? message]) {
    if (_state == state) return;
    _state = state;
    _stateController.add(SocketStateChanged(state: state, errorType: errorType, errorMessage: message));
  }

  SocketErrorType _classifySocketError(SocketException e) {
    final msg = e.message.toLowerCase();
    if (msg.contains('refused') || msg.contains('111')) return SocketErrorType.connectionRefused;
    if (msg.contains('timeout') || msg.contains('110')) return SocketErrorType.timeout;
    if (msg.contains('unreachable') || msg.contains('101') || msg.contains('network')) {
      return SocketErrorType.networkUnreachable;
    }
    if (msg.contains('reset') || msg.contains('104')) return SocketErrorType.connectionReset;
    if (msg.contains('tls') || msg.contains('ssl')) return SocketErrorType.tlsError;
    return SocketErrorType.unknown;
  }
}

class SocketManagerException implements Exception {
  final String message;
  SocketManagerException(this.message);

  @override
  String toString() => 'SocketManagerException: $message';
}

File: Soulseek.Protocol/Connection/SocketManager.cs
using System.Net;
using System.Net.Sockets;

namespace Soulseek.Protocol.Connection;

public enum SocketState
{
    Disconnected,
    Connecting,
    Connected
}

public enum SocketErrorType
{
    None,
    ConnectionRefused,
    DnsFailure,
    Timeout,
    TlsError,
    NetworkUnreachable,
    ConnectionReset,
    Unknown
}

public record SocketStateChanged(SocketState State, SocketErrorType ErrorType = SocketErrorType.None, string? ErrorMessage = null);

public record MessageReceived(Messages.SoulseekMessage Message);

public interface ISocketTransport
{
    SocketState State { get; }
    IObservable<Messages.SoulseekMessage> Messages { get; }
    IObservable<SocketStateChanged> StateChanges { get; }
    Task Connect(string host, int port, TimeSpan? timeout = null);
    Task Disconnect();
    void SendMessage(Messages.SoulseekMessage message);
    void SendRaw(int code, byte[] payload);
    void Dispose();
}

public class SocketManager : ISocketTransport, IDisposable
{
    private TcpClient? _tcpClient;
    private NetworkStream? _stream;
    private SocketState _state = SocketState.Disconnected;
    private CancellationTokenSource? _cts;
    private Task? _readLoopTask;
    private readonly SemaphoreSlim _writeLock = new(1, 1);

    private readonly Subject<Messages.SoulseekMessage> _messageSubject = new();
    private readonly Subject<SocketStateChanged> _stateSubject = new();
    private readonly object _bufferLock = new();
    private readonly List<byte> _buffer = new();

    private const int HeaderSize = 8;

    public SocketState State => _state;

    public IObservable<Messages.SoulseekMessage> Messages => _messageSubject;
    public IObservable<SocketStateChanged> StateChanges => _stateSubject;

    public void Dispose()
    {
        Cleanup();
        _messageSubject.Dispose();
        _stateSubject.Dispose();
    }

    public async Task Connect(string host, int port, TimeSpan? timeout = null)
    {
        if (_state is SocketState.Connecting or SocketState.Connected)
            await Disconnect();

        SetState(SocketState.Connecting);

        try
        {
            var addresses = await Dns.GetHostAddressesAsync(host);
            if (addresses.Length == 0)
            {
                SetState(SocketState.Disconnected, SocketErrorType.DnsFailure, $"Could not resolve host: {host}");
                return;
            }

            var ipv4 = addresses.Where(a => a.AddressFamily == AddressFamily.InterNetwork).ToArray();
            var ipv6 = addresses.Where(a => a.AddressFamily == AddressFamily.InterNetworkV6).ToArray();
            var preferred = ipv4.Length > 0 ? ipv4 : (ipv6.Length > 0 ? ipv6 : addresses);

            _tcpClient = new TcpClient();
            _cts = new CancellationTokenSource();

            using var connectCts = new CancellationTokenSource(timeout ?? TimeSpan.FromSeconds(10));
            try
            {
                await _tcpClient.ConnectAsync(preferred[0], port, connectCts.Token);
            }
            catch (OperationCanceledException)
            {
                SetState(SocketState.Disconnected, SocketErrorType.Timeout, $"Connection timed out after {timeout ?? TimeSpan.FromSeconds(10)}");
                Cleanup();
                return;
            }

            _tcpClient.NoDelay = true;
            _stream = _tcpClient.GetStream();

            SetState(SocketState.Connected);

            _readLoopTask = Task.Run(() => ReadLoop(_cts.Token));
        }
        catch (SocketException e)
        {
            SetState(SocketState.Disconnected, ClassifySocketError(e), e.Message);
            Cleanup();
        }
        catch (Exception e)
        {
            SetState(SocketState.Disconnected, SocketErrorType.Unknown, e.Message);
            Cleanup();
        }
    }

    public void Accept(TcpClient tcpClient)
    {
        _tcpClient = tcpClient;
        _tcpClient.NoDelay = true;
        _stream = _tcpClient.GetStream();
        _cts = new CancellationTokenSource();

        SetState(SocketState.Connected);
        _readLoopTask = Task.Run(() => ReadLoop(_cts.Token));
    }

    public async Task Disconnect()
    {
        _cts?.Cancel();
        _cts?.Dispose();
        _cts = null;

        if (_readLoopTask != null)
        {
            try { await _readLoopTask; } catch { }
            _readLoopTask = null;
        }

        lock (_bufferLock) _buffer.Clear();

        if (_stream != null)
        {
            await _stream.DisposeAsync();
            _stream = null;
        }

        if (_tcpClient != null)
        {
            _tcpClient.Close();
            _tcpClient.Dispose();
            _tcpClient = null;
        }

        SetState(SocketState.Disconnected);
    }

    public void SendMessage(Messages.SoulseekMessage message)
    {
        SendRaw(message.Code, message.Payload);
    }

    public void SendRaw(int code, byte[] payload)
    {
        var stream = _stream;
        if (stream == null || _state != SocketState.Connected)
            throw new SocketManagerException("Not connected");

        var data = Messages.SoulseekMessage.Encode(code, payload);
        _writeLock.Wait();
        try
        {
            stream.Write(data, 0, data.Length);
        }
        finally
        {
            _writeLock.Release();
        }
    }

    private async Task ReadLoop(CancellationToken ct)
    {
        var buffer = new byte[8192];
        try
        {
            while (!ct.IsCancellationRequested)
            {
                var stream = _stream;
                if (stream == null) break;

                var bytesRead = await stream.ReadAsync(buffer, ct);
                if (bytesRead == 0) break;

                lock (_bufferLock)
                {
                    _buffer.AddRange(buffer.AsSpan(0, bytesRead));
                    TryParseMessages();
                }
            }
        }
        catch (OperationCanceledException) { }
        catch (ObjectDisposedException) { }
        catch
        {
            SetState(SocketState.Disconnected, SocketErrorType.ConnectionReset, "Connection lost");
        }
    }

    private void TryParseMessages()
    {
        while (true)
        {
            if (_buffer.Count < HeaderSize) break;

            var totalLength = BitConverter.ToUInt32(_buffer.Take(4).ToArray(), 0);
            var frameSize = 4 + (int)totalLength;

            if (_buffer.Count < frameSize) break;

            var frame = _buffer.Take(frameSize).ToArray();
            _buffer.RemoveRange(0, frameSize);

            try
            {
                var message = Messages.SoulseekMessage.Parse(frame);
                _messageSubject.OnNext(message);
            }
            catch
            {
                // Skip malformed frames
            }
        }
    }

    private void Cleanup()
    {
        _cts?.Cancel();
        _cts?.Dispose();
        _cts = null;

        _writeLock.Wait();
        try
        {
            _stream?.Close();
            _stream?.Dispose();
            _stream = null;

            _tcpClient?.Close();
            _tcpClient?.Dispose();
            _tcpClient = null;
        }
        finally
        {
            _writeLock.Release();
        }

        lock (_bufferLock) _buffer.Clear();
    }

    private void SetState(SocketState state, SocketErrorType errorType = SocketErrorType.None, string? message = null)
    {
        if (_state == state) return;
        _state = state;
        _stateSubject.OnNext(new SocketStateChanged(state, errorType, message));
    }

    private static SocketErrorType ClassifySocketError(SocketException e)
    {
        return e.SocketErrorCode switch
        {
            SocketError.ConnectionRefused => SocketErrorType.ConnectionRefused,
            SocketError.TimedOut => SocketErrorType.Timeout,
            SocketError.NetworkUnreachable => SocketErrorType.NetworkUnreachable,
            SocketError.ConnectionReset => SocketErrorType.ConnectionReset,
            _ => SocketErrorType.Unknown
        };
    }
}

public class SocketManagerException : Exception
{
    public SocketManagerException(string message) : base(message) { }
}